### PR TITLE
uptake from the internal terraform config

### PR DESCRIPTION
**Pull request by guoyongzhang**
Uptake from the internal terraform config to support OCNE 1.7, 1.6, as well as 1.5.  The default OCNE version is 1.7.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,152 +1,165 @@
...
-  master_image_id     = var.master_image_id
-  master_shape        = var.master_shape
-  master_size         = var.master_size
-  master_upgrade      = var.master_package_upgrade
-  ssh_public_key      = ""
-  ssh_public_key_path = var.ssh_public_key_path
-  timezone            = var.master_timezone
-
+module "oci-ocne-vcn" {
+  source     = "./modules/oci-ocne-vcn"
```
Comment (hyder) [inline main.tf]: please use the vcn module https://registry.terraform.io/modules/oracle-terraform-modules/vcn/oci/latest

File: main.tf [hunk 2]
```diff
@@ -1,152 +1,165 @@
...
+  count      = (var.deploy_networking && var.bastion_enabled) || var.bastion_enabled ? 1 : 0
```
Comment (hyder) [inline main.tf]: please rename your variable to imperative form i.e. verb_noun  e.g. enable_bastion

Comment (guoyongzhang) [reply main.tf]: renamed variables:
- bastion_enabled -> enable_bastion
- notification_enabled -> enable_notification

File: main.tf (hunk 1 above)
Comment (guoyongzhang) [reply main.tf]: This module is renamed to oci-ocne-network, which sets up VCN, subnets, as well as, bastion.
I did remove terraform-oci-vcn module and use oracle-terraform-modules/vcn/oci in https://github.com/oracle-terraform-modules/terraform-oci-olcne/blob/58cade472a834306e446e9820642a6482f694281/modules/oci-ocne-network/main.tf#L6

Comment (hyder) [reply main.tf]: cool thank you.

Comment (hyder) [reply main.tf]: I see you are also embedding the bastion module in network. We tried this approach in the oke module and we were using the base module. It caused us endless grief especially when we had to upgrade, add a feature or a fix in the submodule. We had to cut a release every time. 

Can I suggest you keep submodules at 1 level as much a possible?  It's not that it's bad, it'll be easier for you to maintain going forward

File: main.tf (hunk 2 above)
Comment (hyder) [reply main.tf]: thanks

File: main.tf [hunk 3]
```diff
@@ -1,152 +1,165 @@
...
-  worker_shape        = var.worker_shape
-  worker_size         = var.worker_size
-  worker_upgrade      = var.worker_package_upgrade
-  ssh_public_key      = ""
-  ssh_public_key_path = var.ssh_public_key_path
-  timezone            = var.worker_timezone
-
-  # olcne public load balancer to attach to worker pool
-  oci_loadbalancer_id = module.loadbalancer.pub_lb_id
+module "infrastructure" {
```
Comment (hyder) [inline main.tf:71]: What kind of infrastructure does this module handle? Perhaps a short comment will be useful here

File: main.tf [hunk 4]
```diff
@@ -1,152 +1,165 @@
...
-  # workers ip addresses for backend resources
-  olcne_workers = module.worker.worker_ip_list
+module "vault" {
+  source     = "./modules/terraform-oci-vault"
+  count      = var.use_vault ? 1 : 0
+
+  user_id                  = var.user_id
+  region                   = var.region
+  tenancy_id               = var.tenancy_id
+  fingerprint              = var.fingerprint
```
Comment (hyder) [inline main.tf:107]: Why is the provider parameters being passed to the submodule here? If you need to initialize another provider you can do that in the root itself and pass a list of providers. see a more elaborate example here: https://github.com/oracle-terraform-modules/terraform-oci-verrazzano/blob/main/main.tf#L36

Doing so will reduce the number of variables you need to pass between modules

File: main.tf [hunk 5]
```diff
@@ -1,152 +1,165 @@
...
-  secret_id = var.secret_id
-
-  # olcne environment
-  olcne_environment = local.olcne_environment
-
-  helm_version = var.helm_version
-
-  loadbalancer_ip_address = module.loadbalancer.pub_lb_ip
-  deploy_nginx_software = var.deploy_nginx_software
+module "ocne-provision" {
```
Comment (hyder) [inline main.tf:137]: as above, can you please make a short comment about what each module does so we can follow?

File: main.tf [hunk 6]
```diff
@@ -1,152 +1,165 @@
...
+  environment_name          = data.external.ocne_config.result.environment_name
+  kubernetes_name           = data.external.ocne_config.result.kubernetes_name
+  kube_apiserver_port       = var.kube_apiserver_port
+  kube_apiserver_ip         = var.virtual_ip ? module.infrastructure.kube_apiserver_virtual_ip : module.infrastructure.load_balancer_ip
+  virtual_ip                = var.virtual_ip
+  container_registry        = var.container_registry
+  certificate_signing_token = var.use_vault ? module.vault[0].vault_ocne_client_token : ""
+  enable_bastion           = var.enable_bastion
+  bastion_public_ip         = var.enable_bastion ? module.oci-ocne-network[0].bastion_public_ip : var.bastion_public_ip
+  bastion_user              = var.bastion_user
```
Comment (hyder) [inline main.tf]: Do you intend to use other OS e.g. ubuntu for the bastion host? If not, you can keep this to opc

File: modules/oci-ocne-autoscaling/main.tf [hunk 7]
```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2023 Oracle Corporation and/or affiliates.  All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
+
+resource "oci_load_balancer_load_balancer" "lb" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/main.tf:4]: please rename your resources according to their purpose e.g. is this LB the API or a service load balancer

File: modules/oci-ocne-autoscaling/output.tf [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2023 Oracle Corporation and/or affiliates.  All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
+
+output "instances" {
+  description = "Private IP addresses of instances."
+  value       = data.oci_core_instance.instances.*.private_ip
+}
+
+output "load_balancer_ip" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/output.tf:9]: as above, please give meaningful names. outputs are like apis, so they should be clear

File: modules/oci-ocne-autoscaling/variables.tf [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  type        = string
+  description = "The OCID of the compartment in which to create this deployment"
+}
+
+variable "prefix" {
+  type        = string
+  description = "A unique prefix to attach to the name of all cloud resources that are created as a part of this deployment"
+}
+
+variable "instance_shape" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/variables.tf:20]: This should be a map to accomodate flex. see https://github.com/oracle-terraform-modules/terraform-oci-bastion/blob/main/variables.tf#L77

File: modules/oci-ocne-autoscaling/variables.tf [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  type        = string
+  description = "The OCI instance shape to use for all compute resources that are created as part of this deployment"
+}
+
+variable "image_ocid" {
+  type        = string
+  description = "The OCID of the OS image to use when creating all compute resources that are part of this deployment"
+}
+
+variable "subnet_id" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/variables.tf:30]: We should have multiple subnets: control plane, workers, api lb, service lb.

File: modules/oci-ocne-autoscaling/variables.tf [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+  type        = string
+  description = "The OCID of a pre-existing subnet that all newly created cloud resources will be configured to use.  If this variable to set to the empty string, a network configuration will be generated automatically"
+}
+
+// Compute instance specific variables
+variable "ssh_public_key_path" {
+  type        = string
+  description = "The SSH public key path to use when configuring access to any compute resources created as part of this deployment"
+}
+variable "instance_user_data" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/variables.tf:40]: I think this should:
1. have multiple: control plane, workers so you can have different configuration
2. renamed accordingly

File: modules/oci-ocne-autoscaling/variables.tf [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  type        = string
+  description = "Raw cloud-init data that is to be used to instantiate new compute resources"
+}
+
+// Instance pool specific variables
+variable "pool_size" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/variables.tf:46]: are you creating a single pool or multiple pools?

File: modules/oci-ocne-autoscaling/variables.tf [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+  type        = number
+  description = "The initial number of compute instances to spawn as part of this deployment"
+}
+
+
+// Load balancer specific variables
+variable "load_balancer_shape" {
```
Comment (hyder) [inline modules/oci-ocne-autoscaling/variables.tf:52]: like the compute shapes, this should be a map. OCI will be using flexible going forward. See https://github.com/oracle-terraform-modules/terraform-oci-verrazzano/blob/main/variables.tf#L295

File: main.tf (hunk 1 above)
Comment (guoyongzhang) [reply main.tf]: Moved the bastion module from network to https://github.com/oracle-terraform-modules/terraform-oci-olcne/pull/20/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR61

**Review by shihchang — APPROVED** [review]

**Review by scoter-oracle — APPROVED** [review]

**Pull request merged**